File: CSharp/Applications/AvaloniaLoggingDI/App.axaml.cs
using System;
using System.Drawing;
using System.Reflection;
using System.Threading;
using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Data.Core;
using Avalonia.Data.Core.Plugins;
using Avalonia.Markup.Xaml;
using AvaloniaLoggingDI.ViewModels;
using AvaloniaLoggingDI.Views;
using LogViewer.Avalonia;
using MessageBox.Avalonia;
using MessageBox.Avalonia.Enums;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using MsLogger.Core;
using RandomLogging.Service;
using Icon = MessageBox.Avalonia.Enums.Icon;

namespace AvaloniaLoggingDI;

public partial class App : Application
{
    public override void Initialize()
        => AvaloniaXamlLoader.Load(this);

    public override void OnFrameworkInitializationCompleted()
    {
        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            // Line below is needed to remove Avalonia data validation.
            // Without this line you will get duplicate validations from both Avalonia and CT
            ExpressionObserver.DataValidators.RemoveAll(x => x is DataAnnotationsValidationPlugin);

            // catch all unhandled errors
            AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;

            HostApplicationBuilder builder = Host.CreateApplicationBuilder();

            builder
                /*
                 * Note: For information on launch profiles for debugging,
                 *     see article: https://www.codeproject.com/Articles/5354478/NET-App-Settings-Demystified-Csharp-VB
                 */

                // Register the Random Logging Service
                .AddRandomBackgroundService()

                // visual debugging tools
                .AddLogViewer()

                // Microsoft Logger
                //.Logging.AddDefaultDataStoreLogger();

                // uncomment to use custom logging colors (note: System.Drawing namespace)
                //
                .Logging.AddDefaultDataStoreLogger(options =>
                {
                    options.Colors[LogLevel.Trace] = new()
                    {
                        Foreground = Color.White,
                        Background = Color.DarkGray
                    };

                    options.Colors[LogLevel.Debug] = new()
                    {
                        Foreground = Color.White,
                        Background = Color.Gray
                    };

                    options.Colors[LogLevel.Information] = new()
                    {
                        Foreground = Color.White,
                        Background = Color.DodgerBlue
                    };

                    options.Colors[LogLevel.Warning] = new()
                    {
                        Foreground = Color.White,
                        Background = Color.Orchid
                    };
                });

            IServiceCollection services = builder.Services;

            services
                .AddSingleton<MainViewModel>()
                .AddSingleton<MainWindow>(service => new MainWindow
                {
                    DataContext = service.GetRequiredService<MainViewModel>()
                });

            _host = builder.Build();
            _cancellationTokenSource = new();
            
            try
            {
                LogStartingMode();

                // set and show
                desktop.MainWindow = _host.Services.GetRequiredService<MainWindow>();
                desktop.ShutdownRequested += OnShutdownRequested;

                // startup background services
                _ = _host.StartAsync(_cancellationTokenSource.Token);
            }
            catch (OperationCanceledException)
            {
                // skip
            }
            catch (Exception ex)
            {
                ShowMessageBox("Unhandled Error", ex.Message);
                return;
            }
        }

        base.OnFrameworkInitializationCompleted();
    }

    private void OnShutdownRequested(object? sender, ShutdownRequestedEventArgs e)
        => _ = _host!.StopAsync(_cancellationTokenSource!.Token);

    #region Fields

    private IHost? _host;
    private CancellationTokenSource? _cancellationTokenSource;

    #endregion

    private void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
        => ShowMessageBox("Unhandled Error", ((Exception)e.ExceptionObject).Message);

    private void ShowMessageBox(string title, string message)
    {
        MessageBox.Avalonia.BaseWindows.Base.IMsBoxWindow<ButtonResult> messageBoxStandardWindow = MessageBoxManager
            .GetMessageBoxStandardWindow(title, message, ButtonEnum.Ok, Icon.Stop);
 
        messageBoxStandardWindow.Show();
    }

    private void LogStartingMode()
    {
        // Get the Launch mode
        bool isDevelopment = string.Equals(Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES"), "debug",
            StringComparison.InvariantCultureIgnoreCase);

        // initialize a logger & EventId
        ILogger<App> logger = _host!.Services.GetRequiredService<ILogger<App>>();
        EventId eventId = new EventId(id: 0, name: Assembly.GetEntryAssembly()!.GetName().Name);
        
        // log a test pattern for each log level
        logger.TestPattern(eventId: eventId);

        // log that we have started...
        logger.Emit(eventId, LogLevel.Information, $"Running in {(isDevelopment ? "Debug" : "Release")} mode");
    }
}

File: VB/Applications/AvaloniaSerilogVBDI/App.axaml.vb
Imports System.Drawing
Imports System.Reflection
Imports System.Threading
Imports Avalonia
Imports Avalonia.Controls.ApplicationLifetimes
Imports Avalonia.Data.Core
Imports Avalonia.Data.Core.Plugins
Imports Avalonia.Markup.Xaml
Imports CommonVB.Core
Imports LogViewerVB.Avalonia
Imports LogViewerVB.Core
Imports MessageBox.Avalonia
Imports MessageBox.Avalonia.Enums
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Hosting
Imports Microsoft.Extensions.Logging
Imports RandomLoggingVB.Service
Imports Serilog
Imports Serilog.Sinks.LogViewVB.Core

Partial Public Class App : Inherits Application

#Region "Fields"

    Private _host As IHost
    Private _cancellationTokenSource As CancellationTokenSource

#End Region

#Region "Methods"

    Public Overrides Sub Initialize()
        AvaloniaXamlLoader.Load(Me)
    End Sub

    Public Overrides Sub OnFrameworkInitializationCompleted()

        Dim desktop As IClassicDesktopStyleApplicationLifetime =
                TryCast(ApplicationLifetime, IClassicDesktopStyleApplicationLifetime)

        If desktop IsNot Nothing Then

            ' Line below is needed to remove Avalonia data validation.
            ' Without this line you will get duplicate validations from both Avalonia and CT
            ExpressionObserver.DataValidators.RemoveAll(Function(x) TypeOf x Is DataAnnotationsValidationPlugin)

            ' catch all unhandled errors
            AddHandler AppDomain.CurrentDomain.UnhandledException, AddressOf OnUnhandledException

            Dim builder As HostApplicationBuilder = Host.CreateApplicationBuilder()

            ' 
            ' Note: For information on launch profiles for debugging,
            '     see article: https :  //www.codeproject.com/Articles/5354478/NET-App-Settings-Demystified-Csharp-VB
            ' 

            ' Random Logging Service
            builder.AddRandomBackgroundService()

            ' visual debugging tools
            builder.AddLogViewer()

            Dim services As IServiceCollection = builder.Services

            services.AddLogging(
                Sub(cfg)

                    ' Use Default Colors
                    'Log.Logger = New LoggerConfiguration() _
                    '    .ReadFrom.Configuration(builder.Configuration) _
                    '    .WriteTo.DataStoreLoggerSink(
                    '    Function() _host.Services.TryGetService(Of ILogDataStore)) _
                    '.CreateLogger()

                    ' Use Custom Colors
                    Log.Logger = New LoggerConfiguration() _
                            .ReadFrom.Configuration(builder.Configuration) _
                            .WriteTo.DataStoreLoggerSink(
                                Function() _host.Services.TryGetService(Of ILogDataStore),
                                Sub(options)
                                    options.Colors(LogLevel.Trace) = New LogEntryColor() With
                                    {
                                        .Foreground = Color.White,
                                        .Background = Color.DarkGray
                                    }

                                    options.Colors(LogLevel.Debug) = New LogEntryColor() With
                                    {
                                        .Foreground = Color.White,
                                        .Background = Color.Gray
                                    }

                                    options.Colors(LogLevel.Information) = New LogEntryColor() With
                                    {
                                        .Foreground = Color.White,
                                        .Background = Color.DodgerBlue
                                    }

                                    options.Colors(LogLevel.Warning) = New LogEntryColor() With
                                    {
                                        .Foreground = Color.White,
                                        .Background = Color.Orchid
                                    }

                                End Sub) _
                            .CreateLogger()

                    cfg.ClearProviders().AddSerilog(Log.Logger)

                End Sub)

            services _
                .AddSingleton(Of MainViewModel) _
                .AddSingleton(Of MainWindow)(
                    Function(service) New MainWindow() With
                    {
                        .DataContext = service.GetService(Of MainViewModel)
                    })

            _host = builder.Build()
            _cancellationTokenSource = New CancellationTokenSource()

            Try
                LogStartingMode()

                ' set and show
                desktop.MainWindow = _host.Services.GetRequiredService(Of MainWindow)()
                AddHandler desktop.ShutdownRequested, AddressOf OnShutdownRequested

                ' startup background services
                Dim task As Task = _host.StartAsync(_cancellationTokenSource.Token)

            Catch oce As OperationCanceledException

                ' skip

            Catch ex As Exception

                Log.Fatal(ex, "Application terminated unexpectedly")

                ShowMessageBox(ex.Message, "Unhandled Error")

                CleanUp()

            End Try

        End If

        MyBase.OnFrameworkInitializationCompleted()

    End Sub

    Private Sub OnShutdownRequested(sender As Object, e As ShutdownRequestedEventArgs)

        CleanUp()

    End Sub

    Private Sub OnUnhandledException(sender As Object, e As UnhandledExceptionEventArgs)

        Dim exception As Exception = DirectCast(e.ExceptionObject, Exception)

        Log.Fatal(exception, "Application terminated unexpectedly")
        ShowMessageBox(exception.Message, "Unhandled Error")

        CleanUp()

    End Sub

    Private Sub ShowMessageBox(title As String, message As String)

        Dim messageBoxStandardWindow = MessageBoxManager _
                .GetMessageBoxStandardWindow(title, message, ButtonEnum.Ok, Enums.Icon.Stop)

        messageBoxStandardWindow.Show()

    End Sub

    Private Sub LogStartingMode()

        ' Get the Launch mode
        Dim isDevelopment As Boolean = String.Equals(Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES"), "debug",
                                                     StringComparison.InvariantCultureIgnoreCase)


        ' initialize a logger & EventId
        Dim logger As ILogger(Of Application) = _host.Services.GetRequiredService(Of ILogger(Of Application))
        Dim eventId As EventId = New EventId(id:=0, name:=Assembly.GetEntryAssembly.GetName.Name)

        ' log a test pattern for each log level
        logger.TestPattern(eventId)

        ' log that we have started...
        logger.Emit(eventId, LogLevel.Information, $"Running in {If(isDevelopment, "Debug", "Release")} mode")

    End Sub

    Private Sub CleanUp()

        ' tell the background services that we are shutting down
        Dim task As Task = _host.StopAsync(_cancellationTokenSource.Token)

        ' flush logs
        Log.CloseAndFlush()

    End Sub

#End Region

End Class